src/history/hs_rec.c

```diff
@@ -868,9 +870,11 @@ __hs_delete_reinsert_from_pos(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, ui
     WT_UNUSED(key);
 #endif
 
-    /* If we will delete all the updates of the key from the history store, we should not reinsert
-     * any update. */
-    WT_ASSERT(session, ts > WT_TS_NONE || !reinsert);
+    /*
+     * If we will delete all the updates of the key from the history store, we should not reinsert
+     * any update except when the tombstone is not globally visible yet.
+     */
+    WT_ASSERT(session, tombstone_correction || ts > WT_TS_NONE || !reinsert);
```

Comment:
Why we need the extra flag? I think if the tombstone has a ts, then we pass in the ts. If not, we set reinsert to false?

Reply:
In the case of mixed mode, the tombstone doesn't have a timestamp and we may need to retain the older updates for an old reader that is present in the system. So we must reinsert the HS updates with the 0 timestamp instead of removal if it is globally visible. To identify that scenario, the new flag is used.

Reply:
May be we should change the assert to allow reinsert with 0 timestamp instead of adding the flag?

Reply:
Hmm, I did that as a first option and reverted to this as this function is called not only in delete flow, so it is just to make sure that `ts` is not `WT_TS_NONE` when it is called from `insert` flow.

Reply:
The name of the flag is not very clear of your intention.

Reply:
Fixed with a better variable name to indicate an out-of-order tombstone.